### PR TITLE
[Exhibitions] - Reconciles exhibitions to cms exhibitions

### PR DESCRIPTION
**Pull request by sam33rdhakal**
Reconciles the exhibitions using JSON blob's exhibitions.

### CODE REVIEW

**Comment by raigex** [comment]
Looks good to me @agrosner what do you think?

File: db/src/main/kotlin/edu/artic/db/AppDataManager.kt [hunk 1]
```diff
@@ -93,9 +93,15 @@ class AppDataManager @Inject constructor(
 
                         val tours = result.tours
                         if (tours?.isNotEmpty() == true) {
+                            appDatabase.tourDao.clear()
                             appDatabase.tourDao.addTours(tours)
                         }
 
+                        val exhibitionsCMS = result.exhibitions
+                        if (exhibitionsCMS?.isNotEmpty() == true) {
+                            appDatabase.exhibitionCMSDao.addCMSExhibitions(exhibitionsCMS)
```
Comment (agrosner) [inline db/src/main/kotlin/edu/artic/db/AppDataManager.kt]: Do we wanna clear this out before adding? I think soo.

File: db/src/main/kotlin/edu/artic/db/AppDataManager.kt [hunk 2]
```diff
@@ -133,9 +139,25 @@ class AppDataManager @Inject constructor(
...
-                                appDatabase.exhibitionDao.updateExhibitions((result as ArticResult<ArticExhibition>).data)
-
+                                /**
+                                 * Update the sort order of the exhibitions according to the ArticExhibitionCMS
+                                 */
+                                appDatabase.exhibitionCMSDao
+                                        .getAllCMSExhibitions()
+                                        .subscribe {
+                                            @Suppress("UNCHECKED_CAST")
+                                            val list = (result as ArticResult<ArticExhibition>).data
```
Comment (agrosner) [inline db/src/main/kotlin/edu/artic/db/AppDataManager.kt]: this cast is strange. why not cast it further up to the top where its used?

File: db/src/main/kotlin/edu/artic/db/AppDataManager.kt [hunk 3]
```diff
@@ -133,9 +139,25 @@ class AppDataManager @Inject constructor(
...
+                                            val mapExhibitionByID = HashMap<String, ArticExhibition>()
```
Comment (agrosner) [inline db/src/main/kotlin/edu/artic/db/AppDataManager.kt]: don't directly use `HashMap` constructor. use the `mutableMapOf()`

File: db/src/main/kotlin/edu/artic/db/AppDataManager.kt [hunk 4]
```diff
@@ -133,9 +139,25 @@ class AppDataManager @Inject constructor(
...
+
+                                            list.forEach {
```
Comment (agrosner) [inline db/src/main/kotlin/edu/artic/db/AppDataManager.kt]: we can use kotlin collection operators here. just do `list.associateBy { it.id.toString() } `

File: db/src/main/kotlin/edu/artic/db/AppDataManager.kt [hunk 5]
```diff
@@ -133,9 +139,25 @@ class AppDataManager @Inject constructor(
...
+                                                mapExhibitionByID[it.id.toString()] = it
+                                            }
+
+                                            it.forEach {
+                                                mapExhibitionByID[it.id]?.order = it.sort
```
Comment (agrosner) [inline db/src/main/kotlin/edu/artic/db/AppDataManager.kt]: no nested `it`. confusing here. use direct name on the outer object.

File: db/src/main/kotlin/edu/artic/db/daos/ArticExhibitionCMSDao.kt [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import edu.artic.db.models.ArticExhibitionCMS
+import io.reactivex.Flowable
+
+@Dao
+interface ArticExhibitionCMSDao {
+
+    @Query("DELETE FROM ArticExhibitionCMS")
+    fun clear()
+
+    @Query("select * from ArticExhibitionCMS")
```
Comment (agrosner) [inline db/src/main/kotlin/edu/artic/db/daos/ArticExhibitionCMSDao.kt]: one is caps other is lowercase. choose a SQLite syntax pattern and stick with it.

File: db/src/main/kotlin/edu/artic/db/AppDataManager.kt (hunk 2 above)
Comment (raigex) [reply db/src/main/kotlin/edu/artic/db/AppDataManager.kt]: We can cast it there ... but at this time we don't really need to use it anywhere higher, so we cast it when necessary

File: db/src/main/kotlin/edu/artic/db/AppDataManager.kt (hunk 1 above)
Comment (sam33rdhakal) [reply db/src/main/kotlin/edu/artic/db/AppDataManager.kt]: Fixed.

File: db/src/main/kotlin/edu/artic/db/AppDataManager.kt (hunk 3 above)
Comment (sam33rdhakal) [reply db/src/main/kotlin/edu/artic/db/AppDataManager.kt]: Updated.

File: db/src/main/kotlin/edu/artic/db/AppDataManager.kt (hunk 4 above)
Comment (sam33rdhakal) [reply db/src/main/kotlin/edu/artic/db/AppDataManager.kt]: Done.

File: db/src/main/kotlin/edu/artic/db/AppDataManager.kt (hunk 5 above)
Comment (sam33rdhakal) [reply db/src/main/kotlin/edu/artic/db/AppDataManager.kt]: Refactored.

File: db/src/main/kotlin/edu/artic/db/daos/ArticExhibitionCMSDao.kt (hunk 6 above)
Comment (sam33rdhakal) [reply db/src/main/kotlin/edu/artic/db/daos/ArticExhibitionCMSDao.kt]: Updated all of them to lower case.

**Comment by raigex** [comment]
@agrosner changes were made except for 1 (the casting) will merge in and move the casting higher if we use that code at an earlier point

**Pull request merged**